File: implementations/fsharp/tests/Skight.AgentPlatform.FSharp.Tests/GitToolsSpecs.fs
namespace Skight.AgentPlatform.FSharp.Tests

open System
open System.IO
open System.Text.Json
open Expecto
open Skight.AgentPlatform.FSharp

module GitToolsSpecs =

    [<Tests>]
    let tests =
        testList "Git Automation Module Specs" [
            
            testAsync "Feature: Git Status, Commit & Push - Stages and commits changes" {
                let testRepoDir = Path.Combine(Path.GetTempPath(), sprintf "test_git_repo_fs_%s" (Guid.NewGuid().ToString("N")))
                Directory.CreateDirectory(testRepoDir) |> ignore

                try
                    let! _ = TerminalTool.executeCommandDefaultAsync (sprintf "git -C \"%s\" init" testRepoDir)
                    let! _ = TerminalTool.executeCommandDefaultAsync (sprintf "git -C \"%s\" config user.name \"TestUser\"" testRepoDir)
                    let! _ = TerminalTool.executeCommandDefaultAsync (sprintf "git -C \"%s\" config user.email \"test@example.com\"" testRepoDir)

                    let filePath = Path.Combine(testRepoDir, "feature.txt")
                    File.WriteAllText(filePath, "New feature implementation code")

                    let! statusOutput = GitTools.gitStatus testRepoDir |> Async.AwaitTask
                    Expect.stringContains statusOutput "feature.txt" "Git status should report untracked file"

                    let commitArgs = JsonSerializer.Serialize({| message = "Implement new F# feature" |})
                    let! commitOutput = GitTools.gitCommit testRepoDir commitArgs |> Async.AwaitTask
                    Expect.isFalse (commitOutput.StartsWith("Error")) "Commit output should not be error"

                    let! logOutput = TerminalTool.executeCommandDefaultAsync (sprintf "git -C \"%s\" log -1" testRepoDir)
                    Expect.stringContains logOutput "Implement new F# feature" "Git log should record the commit"
                finally
                    if Directory.Exists(testRepoDir) then
                        try Directory.Delete(testRepoDir, true) with _ -> ()
            }
        ]

File: implementations/csharp/tests/Skight.AgentPlatform.MSpec.Tests/GitToolsSpecs.cs
using System;
using System.IO;
using System.Text.Json;
using Machine.Specifications;
using FluentAssertions;
using Skight.AgentPlatform;

namespace Skight.AgentPlatform.MSpec.Tests
{
    [Subject("Git Automation Module - Status, Staging and Committing")]
    public class When_developer_requests_git_commit_on_untracked_files
    {
        Establish context = () =>
        {
            _repoDir = Path.Combine(Path.GetTempPath(), "test_git_repo_mspec_" + Guid.NewGuid().ToString("N"));
            Directory.CreateDirectory(_repoDir);

            TerminalTool.ExecuteCommandAsync($"git -C \"{_repoDir}\" init").GetAwaiter().GetResult();
            TerminalTool.ExecuteCommandAsync($"git -C \"{_repoDir}\" config user.name \"TestUser\"").GetAwaiter().GetResult();
            TerminalTool.ExecuteCommandAsync($"git -C \"{_repoDir}\" config user.email \"test@example.com\"").GetAwaiter().GetResult();

            var filePath = Path.Combine(_repoDir, "feature.txt");
            File.WriteAllText(filePath, "New feature implementation code");
        };

        Because of = () =>
        {
            _statusOutput = GitTools.GitStatusAsync(_repoDir).GetAwaiter().GetResult();
            var commitArgs = JsonSerializer.Serialize(new { message = "Implement new feature" });
            _commitOutput = GitTools.GitCommitAsync(_repoDir, commitArgs).GetAwaiter().GetResult();
            _logOutput = TerminalTool.ExecuteCommandAsync($"git -C \"{_repoDir}\" log -1").GetAwaiter().GetResult();
        };

        It should_report_untracked_files_in_git_status = () =>
            _statusOutput.Should().Contain("feature.txt");

        It should_successfully_commit_changes = () =>
            _commitOutput.Should().NotContain("Error");

        It should_record_the_commit_message_in_git_log = () =>
            _logOutput.Should().Contain("Implement new feature");

        Cleanup after = () =>
        {
            if (Directory.Exists(_repoDir))
            {
                try { Directory.Delete(_repoDir, true); } catch { }
            }
        };

        static string _repoDir;
        static string _statusOutput;
        static string _commitOutput;
        static string _logOutput;
    }
}
